design-proposals/host-ip-pool.md

```diff
@@ -0,0 +1,349 @@
+
+### Implementation Details/Notes/Constraints
+
+#### Day 2
+
+Since kubernetes-nmstate has a distributed architecture each of the handler
+pods will calculate an allacatable ip from the `NodeNetworkIPPool` status and
+retry the process on conflict
+
+Handlers will check NNCEs at Reconcile and if the ip has being already assigned
+it will re-use that and update `NodeNetworkIPPool` status with it, this way the
+`NodeNetworkIPPool` can be reset but IPs are keep.
+
+
+#### Baremetal IPI
+
+The `image-customize-controller` [1] is a golang controller so it would be quite
+easy to import ippool library from kubernetes-nmstate and nmpolicy to render the
+`networkPolicy`.
+
+[1] https://github.com/openshift/image-customization-controller/blob/3ad18911d14dde1c23b660f99bb05e12a21c99e4/pkg/ignition/builder.go#L79
+
+### Risks and Mitigations
+
+Overtime from unexpected failure there are risk of having some IPs allocated
+but not used making the number of available IPs smaller and reaching the end
+sooner than expected, for that reseting the `status.allocated-offsets` should
+reconstruct the pool and bring back those IPs.
+
+### Drawbacks
+
+The baremetal IPI Nmstate do a offline NetworkManager keyfile generation so
+there is no network `currentState` that nmpolicy can reference to, so using
+nmpolicy just to reference the ippools may be too overkill.
+
+If in the future nmstatectl is included at FCOS then the `networkPolicy` will
+be calculated online (but nmpolicyctl will have to be included at FCOS too).
+
+[2] https://github.com/coreos/fedora-coreos-tracker/issues/1175
+
+## Implementation History
+- Generic IPAM mechanism with in memory storage mechanism as either, standalone project, nmpolicy feature or kubernets-nmstate package.
+- IPAM storage mechanism for kubernetes-nmstate use case (day 2)
+- Expression "ippool.[name].allocate" feature either at kubernetes-nmstate or nmpolicy they have to call the created IPAM mechanism
+- Integrate IPAM, nmpolicy and kubernetes-nmstate for simplest use case, allocate an address for the first time.
+- Cover the deallocation scenario where the pool is unreferenced
+- Cover the scale up scenario 
+- Cover the scale down scenario
+- Cover restart scenarios
+- Add `networkPolicy` field at baremetal API [openshift-installer API](https://github.com/openshift/installer/blob/master/pkg/types/baremetal/platform.go#L85)
+- Glue `networkPolicy` field with IPAM + nmpolicy at [image-customization-controller](https://github.com/openshift/image-customization-controller)
+- Test baremetal IPI
+
+## Alternatives
+
+An alternative is to implement directly an IPAM feature at Nmstate with a 
+plugin system for ippool retrieval and allocation storage so at day 2 
+it will use the `NodeNetworkIPPool` and `NodeNetworkConfigurationEnactment` and 
+at baremetal IPI It will read the ippool from the openshift installer config 
+and store the allocations at the nodes file system.
```

Comment:
This means each node would be pre-allocated with its own IP range slice, and would manage it, right ? 

Wouldn't that complicate the scale up / down scenarios ? AFAIU, it would impact each node's already existing range.

Reply:
No, it means that at barmetal IPI the information about allocated IPs per node is store at the filesystem and that this may be converted after starting up into a NNCP that will create the NNCEs but re-using the allocation.